common/src/types.rs

```diff
@@ -0,0 +1,125 @@
+use crate::crypto::*;
+use serde::{Deserialize, Serialize};
+
+pub type VotingPower = u64;
+/// A UNIX timestamp measured in milliseconds.
+pub type Timestamp = u64;
+/// A block height. The genesis block is at height 0.
+pub type BlockHeight = u64;
+pub type ConsensusRound = u64;
+pub type FinalizationProof = Vec<(PublicKey, TypedSignature<BlockHeader>)>;
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct Member {
+    pub public_key: PublicKey,
+    pub name: String,
+    pub governance_voting_power: u64,
+    pub consensus_voting_power: u64,
+    pub governance_delegations: Option<PublicKey>,
+    pub consensus_delegations: Option<PublicKey>,
+    // TODO: add various conditions for each delegation.
+    // - Unlock-Automatically-After-N-Blocks
+    // - Unlock-Automatically-After-T-Seconds
+    // - Unlock-If-The-Delegatee-Is-Not-Active
+    // - Unlock-If-The-Validator-Set-Changes
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct BlockHeader {
+    /// The author of this block.
+    pub author: PublicKey,
+    /// The signature of the previous block.
+    pub prev_block_finalization_proof: FinalizationProof,
+    /// The hash of the previous block.
+    pub previous_hash: Hash256,
+    /// The height of this block.
+    pub height: BlockHeight,
+    /// The timestamp of this block.
+    pub timestamp: Timestamp,
+    /// The hash of aggregation of all the commits for this block.
+    pub commit_hash: Hash256,
+    /// The Merkle root of transactions.
+    pub tx_merkle_root: Hash256,
+    /// The Merkle root of the chat log.
+    pub chat_merkle_root: Hash256,
+    /// The Merkle root of the non-essential state.
+    pub repository_merkle_root: Hash256,
+    /// The effective validator set (delegation-applied) for the next block.
+    ///
+    /// The order here is the consensus leader selection order.
+    pub validator_set: Vec<(PublicKey, VotingPower)>,
+    /// The protocol version that must be used from next block.
+    ///
+    /// It must be a valid semver (e.g., `0.2.3`).
+    pub version: String,
```

Comment:
How about using https://docs.rs/regex/latest/regex/ ?

Reply:
regex is for runtime parsing/matching, not type annotation.

Reply:
Then there is no way else than to put it on the comment? Or maybe error handling later in the function checking the version?

Reply:
We might use `semver` crate but I'm not sure that the dependency is ok to import in WASM. Lets' revisit later.

Reply:
Understood.